File: milestone 2/series4/sde_convweak.m
function sde_convweak
% SDE X_t = f(X_t,t)dt + g(X_t,t)dB_t
%     X_0 = x0; t in [0,T] is solved using the EM method
%
% all paths are computed at the same time
N0=10;                            % number of steps on coarsest level
L=4;                              % number of refinement steps
M=5*10^4;                           % number of samples 
itera=10;
T = 1;                            % final time
x0 = 1;                           % initial condition
f=@(x,t)(-sin(x).*(cos(x)).^3);   % coeffients of the SDE
g=@(x,t)cos(x).^2;
G=@(x) (max(x-1.1,0));            



NL = N0*2^L;
B = brownp(T,NL,M);               % paths of Brownian motion on finest level
YT = zeros(L+1,M);                % values of X_T for h=T/(N*2^l), l=0,...,L, all m
Erro1=zeros(L+1,1);
Erro2=zeros(L+1,1);
tic
for it=1:itera
for l=0:L                         % for N = N0, N0*2, ... , N0*2^L
  N = N0*2^l;
  p = 2^(L-l);                    % p = NL/N
  h = T/N; 
  Y = zeros(N+1,M);
  Y(1,:) = x0;
  for j=1:N                       % perform N steps of E-M method:
    dB = B(1+j*p,:) - B(1+(j-1)*p,:);
    t = j*h;
    x=Y(j,:);
    Y(j+1,:) = x + (-sin(x).*(cos(x)).^3)*h + (cos(x).^2).*dB;
  end
  YT(l+1,:) = Y(N+1,:);           % values of X_T
end
Z = G(YT);                        % compute payoffs


%  weak convergence :
Ym=mean(Z,2);
Ymex = mean(G(atan(randn(1,1e7)*sqrt(T)+tan(x0))));%    sample means for Y
Yme = abs(Ym-Ymex);       %    error in means for weak convergence
Erro1=Erro1+Yme;
Erro2=Erro2+Yme.^2;
end
Erro1=Erro1/itera;
Erro2=(Erro2/itera).^0.5;
Lp = L;                       % plot results for l=0,...,Lp 
                          
hv = T./(N*2.^(0:Lp)');
hL = hv(end); p = hv(1)/hL;

figure(1); loglog(hv,Erro1,'-gx',hv,Erro2,'-ro',hL*[1 p],Erro1(end)*[1 p^1.0],hL*[1 p],Erro2(end)*[1 p^1.0]); 
legend('weak error in L^1', 'weak error in L^2', ' h^{1}',' h^{1}'); grid on
xlabel('step size $h$')
ylabel('error')

% compute convergence rate
p = polyfit(log(hv),log(Erro2),1);
disp('Weak rate of convergence L^2')
disp(p(1));

p = polyfit(log(hv),log(Erro1),1);
disp('Weak rate of convergence L^1')
disp(p(1));

toc

File: milestone 2/series4/sde_convweak.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import linregress
from time import time
from brownp import brownp

def sde_convweak():
    N0 = 10
    L = 4
    M = 5 * 10**4
    itera = 10

    T = 1
    x0 = 1
    f = lambda x, t: -np.sin(x) * (np.cos(x))**3
    g = lambda x, t: np.cos(x)**2
    G = lambda x: np.maximum(x - 1.1, 0)

    NL = N0 * 2**L
    B = brownp(T, NL, M)
    YT = np.zeros((L+1, M))
    Erro1 = np.zeros((L+1, 1))
    Erro2 = np.zeros((L+1, 1))

    start_time = time()

    for it in range(itera):
        for l in range(L+1):
            N = N0 * 2**l
            p = 2**(L-l)
            h = T / N
            Y = np.zeros((N+1, M))
            Y[0, :] = x0

            for j in range(1, N+1):
                dB = B[j*p, :] - B[(j-1)*p, :]
                t = j * h
                x = Y[j-1, :]
                Y[j, :] = x + f(x, t) * h + g(x, t) * dB

            YT[l, :] = Y[N, :]

        Z = G(YT)

        Ym = np.mean(Z, axis=1)
        Ymex = np.mean(G(np.arctan(np.random.randn(1, int(1e7)) * np.sqrt(T) + np.tan(x0))))
        Yme = np.abs(Ym - Ymex)
        Erro1 += Yme.reshape(-1, 1)
        Erro2 += (Yme**2).reshape(-1, 1)

    Erro1 /= itera
    Erro2 = np.sqrt(Erro2 / itera)
    Lp = L

    hv = T / (N * 2 ** np.arange(0, Lp+1))
    hL = hv[-1]
    p = hv[0] / hL

    plt.loglog(hv, Erro1, '-gx', hv, Erro2, '-ro', hL * np.array([1, p]), Erro1[-1] * np.array([1, p**1.0]), hL * np.array([1, p]), Erro2[-1] * np.array([1, p**1.0]))
    plt.legend(['weak error in L^1', 'weak error in L^2', 'h^{1}', 'h^{1}'])
    plt.grid(True)
    plt.xlabel('step size $h$')
    plt.ylabel('error')
    plt.show()

    slope, _, _, _, _ = linregress(np.log(hv), np.log(Erro2.flatten()))
    print('Weak rate of convergence L^2:', slope)

    slope, _, _, _, _ = linregress(np.log(hv), np.log(Erro1.flatten()))
    print('Weak rate of convergence L^1:', slope)

    elapsed_time = time() - start_time
    print('Elapsed time:', elapsed_time)

sde_convweak()
